### ISSUE TITLE
ib.portfolio() returns [] in case of multiple accounts

### ISSUE THREAD

**Issue by klemenkk** [ib-api-reloaded/ib_async#72]
Having multiple (linked) individual accounts does not allow one to call `ib.portfolio()` since it returns `[]` although there are plenty positions in all of the linked accounts.

The only thing that gets an approximation of what I need for me is to call `ib.reqAccountUpdates("All")` and then `ib.sleep(0)`. But the problem is that for instruments where I have simultaneous positions in several accounts the positions are aggregated.

I have also tried calling `ib.reqAccountUpdates("long_account_code") ib.reqAccountUpdates("short_account_code") ib.sleep(0)` and this actually worked for some time but now again stopped working (trust me, it is true) so needed to revert back to calling `ib.reqAccountUpdates("All")`.

I believe there must be a bug in `ib.portfolio()` when dealing with multiple linked individual accounts.

See also [How to get portfolio items for multiple accounts #61](https://github.com/ib-api-reloaded/ib_async/discussions/61)

**Comment by mattsta** [comment]
Multiple accounts are a setup more advanced than we have available for testing currently, so it's not a priority at the moment. Maybe eventually we can get around to it?

For now, this falls back on the old guideline of "this is an uncompensated volunteer project to help a $50 billion corporation for free because their first party interfaces aren't good enough" 🤷‍♂️

Some avenues to look into though:

- those account detail APIs _do_ take account IDs as you mentioned, so it should pass through if you specify multiple accounts manually (one would think)?
- Any attempt to call `ib.sleep(0)` directly somewhat shows a broken system design if people are trying to use it as a hack to "wait for updates." A fully async system never needs to call `sleep(0)` to wait for data updates, plus there is the built-in event notification system too for live notifications. Overall, there's nothing magic about `sleep(0)` and it's more of an anti-pattern if used to intentionally get data results. Also `sleep(0)` fails if it takes more than one event loop cycle to get data back? What if it takes 5 seconds? The event loop exists so we don't manually wait on things.

**Comment by ClimberMel** [comment]
See my comments in https://github.com/ib-api-reloaded/ib_async/discussions/61
I have multiple accounts so I can test this out.  But I would need more detail as to what you want to call portfolio for.
I can call portfolio as I explain in the other issue.  I can also use positions, so I need to know specifics of what you need to retrieve to be able to provide a solution.

**Comment by klemenkk** [comment]
I just need the portfolio() to return the data in the same format as now for one account. Just that it has positions (with realized and unrealized p&l) for all accounts. So something like Account,Symbol,Position,Average Price,Market Price,Market Value,Unrealized P&L,Realized P&L but where "rows"(positions) are from all accounts and are not aggregated (so a row bind of all the positions across all the accounts).

**Comment by JohnFarnea** [comment]
I had the same issue - portfolio() returned [] after creating a sub account in IBKR. It had been working fine prior to the sub account creation for over a year.

I was able to get it working by passing in the account= parameter to the ib.connect() function.

Hope this helps

**Comment by ClimberMel** [comment]
If you use this as a start it will cycle through a list of accounts and return the data for each.  I collect all the data from each and append it to various lists so that I can create various reports and displays with it.
There is a lot more code, but this is the part that connects to each account...

```
for i, a in enumerate(acct):
    myacct = a['account']
    alias = a['alias']
    
    ib = IB()
    ib.connect(pconnect["ip"], pconnect["port"], clientId=pconnect["id"], account=myacct)
    #ib.connect(pconnect["ip"], 7496, clientId=pconnect["id"], account=myacct)

    myPort = ib.portfolio()
```
Just checked and I have the entire code at: https://github.com/ClimberMel/ib_tools/blob/main/code/portfolio.py
You just need to update it from ib_insync to ib_async.  
The ```import code.file as file``` is a module I created to read all my parms from a json file.  Just replace the connect with your own connect string and myacct with a list containing the accounts.
